### PR TITLE
Circ 1058

### PR DESCRIPTION
**Pull request by alexanderkurash**
Resolves [CIRC-1058](https://issues.folio.org/browse/CIRC-1058).

## Purpose
To implement a new approach to block overrides (explained in this [design doc](https://wiki.folio.org/display/DD/Overriding+blocks+in+mod-circulation)) we need to be able to return a full list of validation errors. This will allow users to not spend time trying to override check-out blocks in cases when they won't be able to finish it anyway. 

## Approach
All of the validation errors are handled by the CirculationErrorHandler and accumulated. When validation stage is finished, we return a full list of errors in case if overridable blocks are present.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/test/java/api/loans/CheckOutByBarcodeTests.java
Comment (marcjohnson-kint) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:362]: How can it be determined that an item is not loanable when the borrower cannot be found?

As I understand it, the decision as to whether an item is loanable is based upon the loan policy, which is based upon the circulation rules and those require a patron group.

Comment (alexanderkurash) [reply src/test/java/api/loans/CheckOutByBarcodeTests.java]: I agree, there is no point in checking whether the item is not loanable or item limit is reached without a loan policy, so I've disabled those validations for such case and also removed this piece from the test.

File: src/test/java/api/loans/CheckOutByBarcodeTests.java
Comment (mtraneis) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1338]: Is there any way in a test to simulate the act of overriding the blocks to prove that the checkout action can still proceed or is it out of scope?

File: src/main/java/org/folio/circulation/domain/validation/LoanPolicyValidator.java
Comment (alexanderkurash) [reply src/main/java/org/folio/circulation/domain/validation/LoanPolicyValidator.java]: Removed it. Actually, I've just realized that this validation is done twice. It is a copy of `RegularCheckOutStrategy.refuseWhenItemIsNotLoanable()` which was a part of the `CheckOutStrategy.checkOut()` process. I needed to perform this validation earlier in the chain to accumulate it together with other validation errors and return a failure before the actual check-out, this is why I've extracted it to a separate class. As a result, it's now done twice - first time directly in the check-out chain (it's skipped in case of a OverrideCheckOutStrategy) and second time in the strategy. So I've removed this second validation.

File: src/main/java/org/folio/circulation/resources/handlers/error/CirculationErrorType.java
Comment (alexanderkurash) [reply src/main/java/org/folio/circulation/resources/handlers/error/CirculationErrorType.java]: Added

File: src/test/java/api/loans/CheckOutByBarcodeTests.java
Comment (alexanderkurash) [reply src/test/java/api/loans/CheckOutByBarcodeTests.java:1338]: Sorry for the confusion, I've changed the description in the story. This PR's purpose is to accumulate validation errors to be able to return multiple errors instead of failing with the first one. This is the base for the next check-out override story (CIRC-1061) that will deal with the overrides themselves, so overrides and overridable errors are actually out of scope here and I've removed all mentions of them from this PR.

**Comment by alexanderkurash** [comment]
@marcjohnson-kint Are you planning to continue reviewing this PR or maybe to wait for a review from **folio-org/circulation-backend-code-owners**? We were planning to merge it early this week because there's more override work coming based on this PR.

**Comment by marcjohnson-kint** [comment]
> Are you planning to continue reviewing this PR 

I don't have any more feedback for this pull request. Apologies for not getting back to it sooner.

> or maybe to wait for a review from **folio-org/circulation-backend-code-owners**? 

Why are you waiting for a review from the code owners?

> We were planning to merge it early this week because there's more override work coming based on this PR.

Once it is approved by a member of the Vega team it can be merged.

**Comment by alexanderkurash** [comment]
> Why are you waiting for a review from the code owners?
We're not really waiting, I just saw that you requested a review from this group on Friday, so I thought maybe you wanted some of them to take a look.
> Once it is approved by a member of the Vega team it can be merged.
Got it, thanks.

**Comment by marcjohnson-kint** [comment]
> We're not really waiting, I just saw that you requested a review from this group on Friday, so I thought maybe you wanted some of them to take a look.

Ah, apologies, that's likely a mistake on my part

**Review by mtraneis — COMMENTED** [review]
Because I don't have substantial knowledge in the loan domain I reviewed the code based on the coding practices alone. I think for the most part everything looks all right. There are a couple of areas where I flagged "Validator's order changed, moved up from previous implementation.". I assume that it's okay to rearrange the order somewhat.

File: src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java
Comment (alexanderkurash) [reply src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java:107]: Right, previously it was:
- get **user**
- check that **user** is not blocked
- get **proxy user**
- check that **user** is active
- check that **proxy user** is active
I've moved the 4th check up, so that all user-related checks are grouped together, as well as proxy-user-related checks:
- get **user**
- check that **user** is active
- check that **user** is not blocked
- get **proxy user**
- check that **proxy user** is active
 This validation order makes a little more sense to me, but theoretically this may cause a failure with a different validation error in case when the user is both inactive and blocked. Apparently we don't have a test for this (otherwise the build would fail), probably because it wasn't important to anyone - check-out will be refused anyway. But I can reorder it back if you want to.

File: src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java:176]: Can this method and previous one be combined into one? This way we can avoid these redundant checks.

File: src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java:194]: Same as above.

File: src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java:214]: Same as above.

File: src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java:228]: `requestQueueRepository.get(loanAndRelatedRecords)` does no seem to throw validation errors.

File: src/main/java/org/folio/circulation/resources/handlers/error/CirculationErrorType.java
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/handlers/error/CirculationErrorType.java:44]: Please check if all these errors are used and remove redundant ones.

File: src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java
Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java:104]: If you think it's important to reorder it back to the previous order then please do it, but if you think that the theoretical validation error that may arise when the user is both inactive and blocked is gracefully handled then there is no need to restore the previous validation order.

**Review by OleksandrVidinieiev — APPROVED** [review]

File: src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java
Comment (alexanderkurash) [reply src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java]: Actually, I've reordered it back. It's probably not important at all, but CIRC-1058 doesn't require this change and after all I think it's better to avoid these "while we're here" kind of improvements.

**Pull request merged**
